lib/bolt/node/winrm.rb

```diff
@@ -333,15 +514,27 @@ def _run_task(task, input_method, arguments)
       end
 
       with_remote_file(task) do |remote_path|
-        if powershell_file?(remote_path) && stdin.nil?
-          # NOTE: cannot redirect STDIN to a .ps1 script inside of PowerShell
-          # must create new powershell.exe process like other interpreters
-          # fortunately, using PS with stdin input_method should never happen
-          output = execute("try { &""#{remote_path}"" } catch { exit 1 }")
-        else
-          path, args = *process_from_extension(remote_path)
-          output = execute_process(path, args, stdin)
-        end
+        output =
+          if powershell_file?(remote_path) && stdin.nil?
+            # NOTE: cannot redirect STDIN to a .ps1 script inside of PowerShell
+            # must create new powershell.exe process like other interpreters
+            # fortunately, using PS with stdin input_method should never happen
+            if input_method == 'powershell'
+              execute(<<-PS)
+& { # create a new scope
+  $private:taskArgs = Get-ContentAsJson (
+    $utf8.GetString([System.Convert]::FromBase64String('#{Base64.encode64(JSON.dump(arguments))}'))
+  )
+  try { & '#{remote_path}' @taskArgs } catch { exit 1 }
```

Comment:
On Windows, `"` is a reserved character for filenames, but `'` is allowed, which will break this interpolation - https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396

I'm responsible for the original `&""#{remote_path}""` in https://github.com/puppetlabs/bolt/commit/ea68173084041cf5fddf5f97a2febfce18834dda#diff-0ce90a36bcfd738e65f1aa3cc4892943R330 ... and it appears I did it that way because I wanted `"`, but had to escape as `""` given a Ruby interpolated string.

You mentioned a problem with the quoting here - what were the specifics?

Reply:
```
$ ruby -e 'str = "text"; puts ">>""#{str}""<<"'
>>text<<
```
i.e. it seems to me to be equivalent to:
```ruby
">>#{str}<<"
```

Further I was assuming that the string is passed to a running PowerShell session by some other means than a Windows cmd.exe command line, if that assumption is wrong, then the single quotes cannot be used for the reason you point out.

Reply:
Yeah, you're totally right - I screwed up the initial implementation there.

Original intent was supposed to wrap the path with double quotes in the event it had spaces in it (though it shouldn't). I should have used `\"` instead of `""` (was probably confusing the Ruby interpolation rules with the shell escaping rules!)

```
[1] pry(main)> foo = 'c:\program files\file\test script.ps1'
=> "c:\\program files\\file\\test script.ps1"
[2] pry(main)> "& ""#{foo}"""
=> "& c:\\program files\\file\\test script.ps1"
[3] pry(main)> "& \"#{foo}\""
=> "& \"c:\\program files\\file\\test script.ps1\""
```

Good catch on my bug! I think we still do want to use double quotes rather than single to wrap the path given the Windows filename rules...

Reply:
I misread your comment, sorry. I thought the problem was quoting on the windows command line ...
Now I see your concern is the possibility of having a `'` in a windows filename, in the case the most appropriate quoting would IMO be:
```ruby
%(try { & "#{remote_path}" } catch { exit 1 })
```